### PR TITLE
Implement basic operations in Haskell 

### PR DESCRIPTION
**Pull request by bollu**
Currently, numeric tower works - so `+`, `-`, `/`, `*`, `**` all work.

You can create a symbolic integer by constructing it as one. This will be constructed as a `BasicSym` because it has implemented the `fromInteger` method of the `Num` typeclass. For example, something like this:

```
*Symengine Symengine> let x = -3 :: BasicSym
*Symengine Symengine> x
-3
```

A similar construction works for `float` as well, like so:

```
*Symengine Symengine> let y = 3.5 :: BasicSym
*Symengine Symengine> y
7/2
```

Rationals can also be constructed by using the `rational` function in `Symengine` by passing a numerator and a denominator

```
*Symengine Symengine> rational 2 10
1/5
*Symengine Symengine> let x = rational 1 10
*Symengine Symengine> let y = rational 2 10
*Symengine Symengine> x + y
3/10
*Symengine Symengine> x - y
-1/10
*Symengine Symengine> x * y
1/50
*Symengine Symengine> x / y
1/2
```

complex numbers can be constructed by either calling the `complex` function with a real and imaginary part, or by manually constructing them with the `im` value exported by `Symengine`

```
*Symengine Symengine> let x = complex 1 2
*Symengine Symengine> let y = complex 3 4
*Symengine Symengine> x
1 + 2*I
*Symengine Symengine> y
3 + 4*I
*Symengine Symengine> x + y
4 + 6*I
*Symengine Symengine> x - y
-2 - 2*I
*Symengine Symengine> x * y
-5 + 10*I
*Symengine Symengine> let z = 10 + 20 * im
*Symengine Symengine> z
10 + 20*I
```

Differentiation and symbolics work by first creating a symbol through the `symbol` function. You can then use it just like you would use a SymPy symbol. Differentiation works by calling the `diff` function with an expression and the symbol to differentiate by.

```
*Symengine Symengine> let x = symbol "x"
*Symengine Symengine> x
x
*Symengine Symengine> let y = sin(x)
*Symengine Symengine> y
sin(x)
*Symengine Symengine> diff y x
cos(x)
*Symengine Symengine> let z = symbol "z"
*Symengine Symengine> let y = sin(x * z)
*Symengine Symengine> y
sin(x*z)
*Symengine Symengine> diff y x
z*cos(x*z)
*Symengine Symengine> diff y z
x*cos(x*z)
```

There's stuff like `log` and `logBase` that I've left undefined because I couldn't find the definitions. I guess I'll have to work on the wrappers and bind these.

Next stop: Polynomials and matrices.

### CODE REVIEW

**Comment by bollu** [comment]
@isuruf  - ping. Build doesn't break, and the tests for the trig stuff pass. I'll add tests for the other parts in another PR.

I was wondering, is Codeclimate enabled? It should also be showing up but I don't see it

**Comment by isuruf** [comment]
Looks good.
Codeclimate is enabled, but I don't know why it has stopped running tests.

File: src/Symengine.hs [hunk 1]
```diff
@@ -35,45 +42,240 @@ instance Storable BasicStruct where
...
+instance Fractional BasicSym where
+    (/) = basic_binaryop basic_div_ffi
+    fromRational (num :% denom) = basic_rational_from_integer num denom
+    recip r = one / r
+
+instance Floating BasicSym where
+    pi = Symengine.pi
+    exp x = e ** x
+    log = undefined
+    sqrt x = x  ** 0.5
```
Comment (isuruf) [inline src/Symengine.hs]: Can you change `0.5` to `1/2`?

File: test/Spec.hs [hunk 2]
```diff
@@ -19,11 +21,33 @@ tests = testGroup "Tests" [unitTests]
...
+    , HU.testCase "Basic Trignometric Functions" $
+    do
+      let pi_over_3 = pi / 3 :: BasicSym
+      let pi_over_2 = pi / 2 :: BasicSym
+
+      sin zero @?= zero
+      cos zero @?= one
+      
+      sin (pi / 6) @?= 1 / 2
+      sin (pi / 3) @?= 3 ** 0.5 / 2
```
Comment (isuruf) [inline test/Spec.hs]: Same here

File: test/Spec.hs [hunk 3]
```diff
@@ -19,11 +21,33 @@ tests = testGroup "Tests" [unitTests]
...
+
+      cos (pi / 6) @?= 3 ** 0.5 / 2
```
Comment (isuruf) [inline test/Spec.hs]: Same here

**Comment by isuruf** [comment]
Left a comment, otherwise +1 to merge

**Comment by bollu** [comment]
@isuruf - made the `1/2` change

**Pull request merged**